### ISSUE TITLE
Make the server URL and token configurable.

### ISSUE THREAD

**Issue by NorfairKing** [NorfairKing/hastory#12]
To use the hastory server on the client side, we still need to make some changes.
I think it would be best to pass these thing into the gather script generator, which should then generate the appropriate code to pass in the arguments into `hastory gather`.

**Comment by NorfairKing** [comment]
@yigitozkavci

**Comment by StevenXL** [comment]
Hi @NorfairKing! I *think* this issue is still open.

Currently, [dispatch](https://github.com/NorfairKing/hastory/blob/0af97aeb4219705cf13dae4940797bb526e3eb48/hastory-cli/src/Hastory/Cli.hs#L24) is a computation with access to a value of type [Settings](https://github.com/NorfairKing/hastory/blob/0af97aeb4219705cf13dae4940797bb526e3eb48/hastory-cli/src/Hastory/Cli/OptParse/Types.hs#L58), which will contain the server's info only if the `URL` and `TOKEN` flags are [passed in.](https://github.com/NorfairKing/hastory/blob/0af97aeb4219705cf13dae4940797bb526e3eb48/hastory-cli/src/Hastory/Cli/OptParse.hs#L21)

The interesting problem for me with this issue comes down to testing. How do we test [genGatherWrapperScript](https://github.com/NorfairKing/hastory/blob/0af97aeb4219705cf13dae4940797bb526e3eb48/hastory-cli/src/Hastory/Cli/Commands/GenGatherWrapper.hs#L3), given that the result of that IO action is `()`. 

There are a few options:

**Extract Effect into a Typeclass**

We could [extract the effect](https://www.fpcomplete.com/blog/2017/06/readert-design-pattern#regain-purity) into a typeclass. For the IO monad, we'd print it to stdout, and we'd have a different monad in testing that would write the output to a log. In pseudo-code:

```haskell
class Monad m => Display m where
   display :: String -> m ()

instance Display IO where
   display = putStr

instance Display (StackWithWriter) where
  display = tell
```

Another option is parameterized `genGatherWrapperScript` so that it takes in the handle to write to: `genGatherWrapperScript :: Handle -> IO ()`.

Finally, we can try to separate the pure and impure code in `genGatherWrapperScript`, and test only the pure code:

```haskell
genGatherWrapperScript :: IO ()
genGatherWrapperScript =
  userInput <- gatherUserInput
 let script = scriptFromUserInput userInput
  putStrLn script

script :: UserInput -> String
```

In the pseudo-code above, we'd only test `script` and not `genGatherWrapperScript`. 

What are your thoughts on these different approaches?

**Comment by NorfairKing** [comment]
> Extract Effect into a Typeclass

Absolutely not.
I don't believe in mocking.

> Finally, we can try to separate the pure and impure code in genGatherWrapperScript, and test only the pure code:

This is good enough for me.

In particular, we want to make sure we can use both a `--comand-line-flag` as well as an `ENVIRONMENT_VARIABLE`. See https://github.com/NorfairKing/intray/blob/master/intray-server/src/Intray/Server/OptParse.hs and https://github.com/NorfairKing/intray/blob/master/intray-server/src/Intray/Server/OptParse/Types.hs for the standard way to do this.

**Comment by StevenXL** [comment]
Perfect. Thanks! This should give me enough to create a PR that we can refine further.

**Comment by StevenXL** [comment]
Closed by #27?

**Comment by NorfairKing** [comment]
Yess
